dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/impl/AggregateContext.kt

```diff
@@ -0,0 +1,135 @@
+            val path = stack.currentPath()
+            check(!toBeSent.messages.containsKey(path)) {
+                """
+                    Alignment was broken by multiple aligned calls with the same path: $path.
+                    The most likely cause is an aggregate function call within a loop
+                """.trimIndent()
+            }
+            toBeSent = toBeSent.copy(messages = toBeSent.messages + (stack.currentPath() to message))
+            state = state + (stack.currentPath() to field.localValue)
+        }
+    }
+
+    override fun <Init, Ret> exchanging(
+        initial: Init,
+        body: YieldingContext<Field<Init>, Field<Ret>>.(Field<Init>) -> YieldingResult<Field<Init>, Field<Ret>>,
+    ): Field<Ret> {
+        val messages = messagesAt<Init>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        val context = YieldingContext<Field<Init>, Field<Ret>>()
+        var res: Option<YieldingResult<Field<Init>, Field<Ret>>>
+        body(context, subject).also {
+            res = it.some()
+            state = state + (stack.currentPath() to it.toReturn.localValue)
+        }
+        return res.getOrElse { error("This error should never be thrown") }.toReturn
+    }
+
+    override fun <Initial, Return> repeating(
+        initial: Initial,
+        transform: YieldingContext<Initial, Return>.(Initial) -> YieldingResult<Initial, Return>,
+    ): Return {
+        val context = YieldingContext<Initial, Return>()
+        var res: Option<YieldingResult<Initial, Return>>
+        transform(context, stateAt(stack.currentPath(), initial)).also {
+            res = it.some()
+            state = state + (stack.currentPath() to it.toReturn)
+        }
+        return res.getOrElse { error("This error should never be thrown") }.toReturn
+    }
+
+    override fun <Initial> repeat(
+        initial: Initial,
+        transform: (Initial) -> Initial,
+    ): Initial =
+        repeating(initial) {
+            val res = transform(it)
+            YieldingResult(res, res)
+        }
+
+    override fun <R> alignedOn(pivot: Any?, body: () -> R): R {
+        stack.alignRaw(pivot)
+        return body().also { stack.dealign() }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun <T> messagesAt(path: Path): Map<ID, T> = messages
+        .filter { it.messages.containsKey(path) }
+        .associate { it.senderId to it.messages[path] as T }
+
+    private fun <T> stateAt(path: Path, default: T): T = previousState.getTyped(path, default)
+}
+
+/**
+ * Projects the field to be aligned with the neighbours.
+ * A field de-alignment can occur when the field is used inside a body of a branch condition.
+ * Take the [field] to project and returns a new field aligned with the neighbours.
+ * This method is meant to be used internally by the compiler plugin.
+ */
+fun <T> Aggregate.project(field: Field<T>): Field<T> {
```

Comment:
I think it should be `inline`

Reply:
Or, actually, no. There's no reason to.

Reply:
I don't see any reason to make it `inline`, it should not produce a performance penalty.
I think make it `inline` implies a change in the application of it on compiler plugin side

Reply:
No yeah, please don't, ignore this conversation and carry on